File: signal-chat/crypto/x3dh.py
"""
X3DH (Extended Triple Diffie-Hellman) key agreement.
Uses dh.py for all DH operations; cryptography lib only for HKDF.
"""
import os
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes
from .dh import DHKeyPair


_INFO = b"UIC-Signal-X3DH-v1"
_F = b"\xff" * 32  # X3DH spec padding


def _hkdf(ikm: bytes, length: int = 32, salt: bytes = b"\x00" * 32) -> bytes:
    return HKDF(
        algorithm=hashes.SHA256(),
        length=length,
        salt=salt,
        info=_INFO,
    ).derive(ikm)


class X3DHBundle:
    """A full prekey bundle for a single user."""

    def __init__(self):
        self.ik = DHKeyPair()          # identity key (long-term)
        self.spk = DHKeyPair()         # signed prekey
        self.opk = DHKeyPair()         # one-time prekey

    def public_bundle(self) -> dict:
        return {
            "ik_pub": self.ik.public_key,
            "spk_pub": self.spk.public_key,
            "opk_pub": self.opk.public_key,
        }


def x3dh_sender(
    sender_ik: DHKeyPair,
    recipient_bundle: dict,
) -> tuple[bytes, int, int, int]:
    """
    Sender-side X3DH.
    Returns (shared_key_32_bytes, ek_pub, sender_ik_pub, opk_pub_used).
    """
    ek = DHKeyPair()  # ephemeral key

    rik = recipient_bundle["ik_pub"]
    rspk = recipient_bundle["spk_pub"]
    ropk = recipient_bundle["opk_pub"]

    dh1 = sender_ik.dh(rspk)
    dh2 = ek.dh(rik)
    dh3 = ek.dh(rspk)
    dh4 = ek.dh(ropk)

    km = _F + dh1 + dh2 + dh3 + dh4
    sk = _hkdf(km)
    return sk, ek.public_key, sender_ik.public_key, ropk


def x3dh_receiver(
    receiver_bundle: X3DHBundle,
    sender_ik_pub: int,
    ek_pub: int,
) -> bytes:
    """
    Receiver-side X3DH — reconstructs the same shared key.
    """
    dh1 = receiver_bundle.spk.dh(sender_ik_pub)
    dh2 = receiver_bundle.ik.dh(ek_pub)
    dh3 = receiver_bundle.spk.dh(ek_pub)
    dh4 = receiver_bundle.opk.dh(ek_pub)

    km = _F + dh1 + dh2 + dh3 + dh4
    return _hkdf(km)

File: signal-chat/frontend/src/crypto/x3dh.js
// X3DH (Extended Triple Diffie-Hellman) — browser side.
// Mirrors crypto/x3dh.py exactly. All DH ops use dh.js (BigInt).
// Only HKDF comes from WebCrypto (same rule as Python: only AES-GCM + HKDF allowed).

import { DHKeyPair, hexToBigInt, bigIntToBytes } from "./dh.js";

const X3DH_INFO = new TextEncoder().encode("UIC-Signal-X3DH-v1");
const F = new Uint8Array(32).fill(0xff);  // same padding as Python

function concat(...arrays) {
  const total = arrays.reduce((s, a) => s + a.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const a of arrays) { out.set(a, offset); offset += a.length; }
  return out;
}

export async function hkdf(ikm, salt = new Uint8Array(32), info = X3DH_INFO, length = 32) {
  const key = await crypto.subtle.importKey("raw", ikm, { name: "HKDF" }, false, ["deriveBits"]);
  const bits = await crypto.subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt, info },
    key, length * 8
  );
  return new Uint8Array(bits);
}

// Sender-side X3DH (Alice's browser)
// senderIK      — DHKeyPair (Alice's identity key, private in localStorage)
// recipientBundle — { ik_pub, spk_pub, opk_pub } hex strings from server
// Returns { sk, ekPub (BigInt), senderIKPub (BigInt) }
export async function x3dhSender(senderIK, recipientBundle) {
  const ek = new DHKeyPair();  // ephemeral key, generated fresh each session

  const rik  = hexToBigInt(recipientBundle.ik_pub);
  const rspk = hexToBigInt(recipientBundle.spk_pub);

  if (!recipientBundle.opk_pub) throw new Error(`No one-time prekey available for this user`);
  const ropk = hexToBigInt(recipientBundle.opk_pub);

  const dh1 = senderIK.dh(rspk);   // DH(sender_IK,   recipient_SPK)
  const dh2 = ek.dh(rik);           // DH(sender_EK,   recipient_IK)
  const dh3 = ek.dh(rspk);          // DH(sender_EK,   recipient_SPK)
  const dh4 = ek.dh(ropk);          // DH(sender_EK,   recipient_OPK)

  const km = concat(F, dh1, dh2, dh3, dh4);
  const sk = await hkdf(km);

  return { sk, ekPub: ek.publicKey, senderIKPub: senderIK.publicKey };
}

// Receiver-side X3DH (Bob's browser)
// receiverKeys  — { ik, spk, opk } DHKeyPairs loaded from localStorage
// senderIKPub   — hex string from session_init header
// ekPub         — hex string from session_init header
// Returns sk (Uint8Array, 32 bytes)
export async function x3dhReceiver(receiverKeys, senderIKPubHex, ekPubHex) {
  const sik = hexToBigInt(senderIKPubHex);
  const ek  = hexToBigInt(ekPubHex);

  const dh1 = receiverKeys.spk.dh(sik);   // DH(recipient_SPK, sender_IK)
  const dh2 = receiverKeys.ik.dh(ek);      // DH(recipient_IK,  sender_EK)
  const dh3 = receiverKeys.spk.dh(ek);     // DH(recipient_SPK, sender_EK)
  const dh4 = receiverKeys.opk.dh(ek);     // DH(recipient_OPK, sender_EK)

  const km = concat(F, dh1, dh2, dh3, dh4);
  return await hkdf(km);
}
